`invdes` validators

yaugenst-flex opened the pull request:
Adds a bunch of validators for `invdes` as well as constraints on some fields.
Depends on #1988

tylerflex commented on tidy3d/plugins/autograd/functions.py:
```diff
@@ -241,7 +241,10 @@ def pad(
-        raise NotImplementedError("Padding larger than the input size is not supported.")
+        raise NotImplementedError(
+            "Padding larger than the input size is not supported. "
+            "The convolution kernel is probably too large for the input."
```
can we display the relevant shapes here and maybe add more instruction on how to fix it?

yaugenst-flex replied:
done

tylerflex commented on tidy3d/plugins/expressions/metrics.py:
```diff
@@ -66,9 +66,9 @@ def evaluate(self, *args: Any, **kwargs: Any) -> NumberType:
             data[self.monitor_name]
             .amps.sel(direction=self.direction, mode_index=self.mode_index)
-            .isel(f=0)
+            .sel(f=self.freqs[0], method="nearest")
```
I still dont quite understand the use of `0` here, doesn't this just ignore the frequency defined in the class?

yaugenst-flex replied:
sure yeah we can use `self.freqs`. it's guaranteed to be a single frequency through the validator

yaugenst-flex replied:
fixed

tylerflex replied:
if self.freqs is a single frequency only maybe it should just be `f: float = None` (and if None we isel?)

yaugenst replied:
not sure i understand? multifrequency support will be added later. currently there is only a validator that checks that len(freqs)==1, which we can remove once multifrequency works

yaugenst-flex replied:
or do you mean making freqs optional and then isel if only a single freq is found in the monitor? but that'll be a pain to validate

tylerflex replied:
No, I think what I'm confused about is that `ModePower.freqs` takes a list of freqs, but it only selects the first one. Why not just make it take a single frequency and call it `f`?

tylerflex replied:
this would make it consistent with the xarray coords too. I think this is basically something I missed when reviewing the original metrics PR

yaugenst-flex replied:
because we do want to support multiple freqs in the feature. it‘s just disabled right now

tylerflex replied:
but in xarray we also would just overload the singular coordinate with a list, eg `f=[f1, f2]`. So maybe we should follow this approach here?

tylerflex replied:
basically might be nice to be able to essentially copy the kwargs passed to `sel(**kwargs)` into `ModePower(**kwargs)`?

yaugenst-flex replied:
ah, i guess i was trying to resemble the modemonitor interface more than the dataarray one with the metrics

tylerflex replied:
yea I think it should probably mirror the ModeAmpsDataArray, because the ModeMontior doesn't do `mode_index` for example? and this is essentially a wrapper for picking data out of the ModeAMpsDataARray right?

yaugenst-flex replied:
yeah that's a good point. I think `**kwargs` will not be possible because we do need to validate all the fields, the `sel`s in the metrics won't be arbitrary. but we can definitely change the naming. might be better to do it in a separate PR though?

tylerflex replied:
no preference on the PR, but best to change before 2.7.5 I think

tylerflex commented on tidy3d/plugins/invdes/design.py:
```diff
@@ -119,6 +122,54 @@ class InverseDesign(AbstractInverseDesign):
 
     _check_sim_pixel_size = check_pixel_size("simulation")
 
+    @pd.root_validator(pre=False)
```
maybe could do this using the post init validators defined in Tidy3dBaseModel? would avoid needing to do values.get() stuff\

The basic idea is to overwrite this method and call your validation methods inside of it
https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/components/base.py#L150

yaugenst-flex replied:
sure. I was thinking it'd be better to validate as early as possible, and here this works, i.e. you don't need an instance of the class to check the validation. I can change it

yaugenst-flex replied:
there are some other instance in this PR where that approach doesn't work and I use the post init validator

tylerflex replied:
ok, I'll leave it up to you, just letting you know so you know about it.

tylerflex commented on tidy3d/plugins/invdes/initialization.py:
```diff
@@ -29,17 +32,34 @@ class RandomInitializationSpec(AbstractInitializationSpec):
+        maxval = values.get("max_value")
+        if minval > maxval:
+            raise ValidationError(
+                f"'max_value' ({maxval}) must be greater or equal than 'min_value' ({minval})"
```
if they are equal, it might lead to some weird behavior in the projections? might want to verify

yaugenst-flex replied:
it's fine, it's essentially the same as uniform initialization

tylerflex commented on tidy3d/plugins/invdes/region.py:
```diff
@@ -159,12 +175,69 @@ class TopologyDesignRegion(DesignRegion):
+            x = self.initial_parameters
+            self.eps_values(x)
+        except Exception as e:
+            raise ValidationError(f"Could not evaluate transformations: {str(e)}") from e
```
will this print `e` twice though?

yaugenst-flex replied:
no

yaugenst-flex replied:
or, more specifically, the validation error will appear only once pretty printed in the logs, but the complete stack trace is preserved. in the stack trace, both ValidationError and the previous exception will show up

yaugenst-flex replied:
we can also `raise from None` instead of `raise from e`, but then the traceback is lost, which I think would be bad for debugging?

tylerflex replied:
I think how you have it is fine, I just get confused about exception handling.

tylerflex reviewed (changes requested):
looks good, thanks. minor suggestions

tylerflex commented:
also note seemes that #1988 is included in this?

yaugenst-flex commented:
> also note seemes that https://github.com/flexcompute/tidy3d/pull/1988 is included in this?

it is because this is branched off of that but has `develop` as a target. will be gone when that PR is merged and this is one rebased

e-g-melo reviewed (commented):
Thanks @yaugenst-flex! The new validators seem very detailed. Tomorrow, I will checkout your branch to run a real example.

e-g-melo commented:
@yaugenst-flex, I ran the [inverse design plugin example](https://docs.flexcompute.com/projects/tidy3d/en/v2.7.4/notebooks/InverseDesign.html) and got these errors in `design_region.to_structure()` and `inv_des.to_simulation()` when trying to call them without `params0`.

![image](https://github.com/user-attachments/assets/45446669-d7d7-41e8-8266-d93e9deeddfd)

Should we make `param0` optional in `design_region.to_structure()`, `inv_des.to_simulation()`, and `inv_des.to_simulation_data()`?

momchil-flex commented on tidy3d/plugins/invdes/region.py:
```diff
@@ -159,12 +175,69 @@ class TopologyDesignRegion(DesignRegion):
+
+    def _validate_gradients(self):
+        """Validate the gradients of the penalties and transformations."""
+        import warnings
```
Should this just be top level?

yaugenst-flex replied:
sure can do. I guess I thought this is only used in the validator to catch an autograd warning, but maybe makes no sense to lazy import

yaugenst-flex commented:
> Should we make param0 optional in design_region.to_structure(), inv_des.to_simulation(), and inv_des.to_simulation_data()?

hm good point, although it would only work for the initial parameters right? so if you want to give e.g. the optimized parameters, you still need to pass them explicitly. might be a bit confusing?

e-g-melo commented:
> hm good point, although it would only work for the initial parameters right? so if you want to give e.g. the optimized parameters, you still need to pass them explicitly. might be a bit confusing?

Ok.. in that case, maybe we can simply modify the example and use `design_region.to_structure(design_region.initial_parameters)`

e-g-melo commented:
@yaugenst-flex, I made the filter radius larger than the `z` direction and set `uniform=(False, False, False)`. Then, I got this validation error. Is it the expected validation error? If that is the case, would it be possible to include a message that explicitly informs the user that the filter radius is larger than the design region in the `z` direction?
![image](https://github.com/user-attachments/assets/0406457a-7962-488f-bdd8-9e0ef6330b03)

e-g-melo commented:
In the case we set `uniform=(False, False, True)` and set the projection filter radius larger than the `x, y` dimensions, this validation error is shown. Would it be possible to have a message that tells the user which component to change to solve the issue? For example:

ERROR: Could not evaluate transformations: **Conic filter radius larger than the 
             design region is not supported**. Padding widths: [240 240], Array      
             shape: (240, 252). Ensure that the conic filter radius is smaller than   
             the corresponding dimension of the design region.

![image](https://github.com/user-attachments/assets/84daf78c-49ac-44df-a3e5-02277265bbc5)

e-g-melo commented:
I got this error in the 2nd iteration step when trying multiple frequencies in the metrics monitor but selecting a single one in metrics. Is it expected to work on this PR?

```py
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
Cell In[27], [line 1](vscode-notebook-cell:?execution_count=27&line=1)
----> [1](vscode-notebook-cell:?execution_count=27&line=1) result = optimizer.run()

File /mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:132, in AbstractOptimizer.run(self, post_process_fn, callback, params0)
    [117](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:117) """Run this inverse design problem from an optional initial set of parameters.
    [118](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:118) 
    [119](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:119) Parameters
   (...)
    [129](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:129)     Deprecated. Initial set of parameters. Use ``TopologyDesignRegion.intialization_spec`` instead.
    [130](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:130) """
    [131](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:131) starting_result = self.initialize_result(params0)
--> [132](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:132) return self.continue_run(
    [133](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:133)     result=starting_result,
    [134](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:134)     num_steps=self.num_steps,
    [135](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:135)     post_process_fn=post_process_fn,
    [136](https://vscode-remote
[InverseDesign_Validators3D.zip](https://github.com/user-attachments/files/17245123/InverseDesign_Validators3D.zip)
+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:136)     callback=callback,
    [137](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:137) )

File /mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:180, in AbstractOptimizer.continue_run(self, result, num_steps, post_process_fn, callback)
    [178](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:178) for step_index in range(done_steps, done_steps + num_steps):
    [179](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:179)     aux_data = {}
--> [180](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:180)     val, grad = val_and_grad_fn(params, aux_data=aux_data)
    [182](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:182)     if anp.allclose(grad, 0.0):
    [183](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:183)         td.log.warning(
    [184](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:184)             "All elements of the gradient are almost zero. This likely indicates "
    [185](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:185)             "a problem with the optimization set up. This can occur if the symmetry of the "
   (...)
    [193](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:193)             "function are contributing to the return value."
    [194](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/plugins/invdes/optimizer.py:194)         )

File ~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/wrap_util.py:20, in unary_to_nary.<locals>.nary_operator.<locals>.nary_f(*args, **kwargs)
     [18](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/wrap_util.py:18) else:
     [19](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/wrap_util.py:19)     x = tuple(args[i] for i in argnum)
---> [20](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/wrap_util.py:20) return unary_operator(unary_f, x, *nary_op_args, **nary_op_kwargs)

File ~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/differential_operators.py:143, in value_and_grad(fun, x)
    [139](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/differential_operators.py:139) if not vspace(ans).size == 1:
    [140](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/differential_operators.py:140)     raise TypeError("value_and_grad only applies to real scalar-output "
    [141](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/differential_operators.py:141)                     "functions. Try jacobian, elementwise_grad or "
    [142](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/differential_operators.py:142)                     "holomorphic_grad.")
--> [143](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/differential_operators.py:143) return ans, vjp(vspace(ans).ones())

File ~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:14, in make_vjp.<locals>.vjp(g)
---> [14](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:14) def vjp(g): return backward_pass(g, end_node)

File ~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:21, in backward_pass(g, end_node)
     [19](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:19) for node in toposort(end_node):
     [20](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:20)     outgrad = outgrads.pop(node)
---> [21](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:21)     ingrads = node.vjp(outgrad[0])
     [22](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:22)     for parent, ingrad in zip(node.parents, ingrads):
     [23](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:23)         outgrads[parent] = add_outgrads(outgrads.get(parent), ingrad)

File ~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:67, in defvjp.<locals>.vjp_argnums.<locals>.<lambda>(g)
     [64](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:64)         raise NotImplementedError(
     [65](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:65)             "VJP of {} wrt argnum 0 not defined".format(fun.__name__))
     [66](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:66)     vjp = vjpfun(ans, *args, **kwargs)
---> [67](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:67)     return lambda g: (vjp(g),)
     [68](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:68) elif L == 2:
     [69](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/autograd/core.py:69)     argnum_0, argnum_1 = argnums

File /mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:594, in _run_bwd.<locals>.vjp(data_fields_vjp)
    [591](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:591) def vjp(data_fields_vjp: AutogradFieldMap) -> AutogradFieldMap:
    [592](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:592)     """dJ/d{sim.traced_fields()} as a function of Function of dJ/d{data.traced_fields()}"""
--> [594](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:594)     sim_adj, adjoint_source_info = setup_adj(
    [595](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:595)         data_fields_vjp=data_fields_vjp,
    [596](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:596)         sim_data_orig=sim_data_orig,
    [597](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:597)         sim_fields_keys=sim_fields_keys,
    [598](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:598)     )
    [600](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:600)     # run adjoint simulation
    [601](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:601)     task_name_adj = str(task_name) + "_adjoint"

File /mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:748, in setup_adj(data_fields_vjp, sim_data_orig, sim_fields_keys)
    [745](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:745) data_vjp_paths = set(data_fields_vjp.keys())
    [747](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:747) num_monitors = len(sim_data_orig.simulation.monitors)
--> [748](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:748) adjoint_monitors = sim_data_orig.simulation.with_adjoint_monitors(sim_fields_keys).monitors[
    [749](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:749)     num_monitors:
    [750](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:750) ]
    [752](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:752) sim_adj, adjoint_source_info = sim_data_vjp.make_adjoint_sim(
    [753](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:753)     data_vjp_paths=data_vjp_paths, adjoint_monitors=adjoint_monitors
    [754](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:754) )
    [756](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/web/api/autograd/autograd.py:756) if _INSPECT_ADJOINT_FIELDS:

File /mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/simulation.py:3633, in Simulation.with_adjoint_monitors(self, sim_fields_keys)
   [3631](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/simulation.py:3631) mnts_fld, mnts_eps = self.make_adjoint_monitors(structure_indices=structure_indices)
   [3632](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/simulation.py:3632) monitors = list(self.monitors) + list(mnts_fld) + list(mnts_eps)
-> [3633](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/simulation.py:3633) return self.copy(update=dict(monitors=monitors))

File /mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:223, in Tidy3dBaseModel.copy(self, deep, **kwargs)
    [221](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:221) kwargs.update(deep=deep)
    [222](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:222) new_copy = pydantic.BaseModel.copy(self, **kwargs)
--> [223](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:223) return self.validate(new_copy.dict())

File ~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:711, in BaseModel.validate(cls, value)
    [708](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:708) value = cls._enforce_dict_if_root(value)
    [710](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:710) if isinstance(value, dict):
--> [711](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:711)     return cls(**value)
    [712](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:712) elif cls.__config__.orm_mode:
    [713](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:713)     return cls.from_orm(value)

File /mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:146, in Tidy3dBaseModel.__init__(self, **kwargs)
    [144](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:144) """Init method, includes post-init validators."""
    [145](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:145) log.begin_capture()
--> [146](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:146) super().__init__(**kwargs)
    [147](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:147) self._post_init_validators()
    [148](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev/tidy3d/tidy3d/components/base.py:148) log.end_capture(self)

File ~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:341, in BaseModel.__init__(__pydantic_self__, **data)
    [339](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:339) values, fields_set, validation_error = validate_model(__pydantic_self__.__class__, data)
    [340](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:340) if validation_error:
--> [341](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:341)     raise validation_error
    [342](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:342) try:
    [343](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/mnt/d/Tidy3D_Dev_Test/test_serializable_ops/~/miniconda3/envs/tidy3d_dev/lib/python3.12/site-packages/pydantic/v1/main.py:343)     object_setattr(__pydantic_self__, '__dict__', values)

ValidationError: 1 validation error for Simulation
monitors
  'monitors' names are not unique, given ['MNT_0', 'MNT_1', 'MNT_2', 'adjoint_fld_4', 'adjoint_fld_4', 'adjoint_eps_4']. (type=value_error.setup)
```

e-g-melo commented:
If I set `length_scale` to -1 I get this error, which is correct.

![image](https://github.com/user-attachments/assets/c931528b-d07a-44b2-9b81-8a9585de643f)

But if I set `length_scale` to 0, I get this other error. What should happen if `length_scale=0`? Should we validate it to `>0` as in the conic filter radius?

![image](https://github.com/user-attachments/assets/a8badc04-7608-46a0-b8d0-7ae31bd26b87)

e-g-melo commented:
`override_structure_dl=0` passed the validation

<img width="797" alt="image" src="https://github.com/user-attachments/assets/6e275727-febf-4a84-9b35-9ad94fb65ae3">

e-g-melo commented:
`learning_rate=0` passed the validation. Should we validate it for `>0`?

<img width="462" alt="image" src="https://github.com/user-attachments/assets/b51bf473-c9a3-408e-bcde-bacad2e7727b">

yaugenst-flex commented:
Hey @e-g-melo, these are great! Thanks a lot for catching all of this. Point by point:

> I made the filter radius larger than the z direction and set `uniform=(False, False, False)`. Then, I got this validation error. Is it the expected validation error?

No this is not expected, it looks like the parameters have the wrong shape. I'll have to look into this one.

> In the case we set `uniform=(False, False, True)` and set the projection filter radius larger than the x, y dimensions, this validation error is shown. Would it be possible to have a message that tells the user which component to change to solve the issue?

Ok I'll try to improve this. It's a bit tricky because scalar dimensions are removed before convolution, and the error is raised in the convolution, so there the information about the physical axis is lost. Maybe validation can happen a bit earlier.

> I got this error in the 2nd iteration step when trying multiple frequencies in the metrics monitor but selecting a single one in metrics. Is it expected to work on this PR?

In principle this should work. It looks like the `adjoint_fld_4` monitor is added twice, which causes simulation validation to fail. So I don't think this is a problem with the newly introduced validation. I'll investigate.

> But if I set `length_scale` to 0, I get this other error. What should happen if `length_scale=0`? Should we validate it to >0 as in the conic filter radius?

Good catch, fixed. `length_scale` must be positive.

> `override_structure_dl=0` passed the validation

Oooh this is interesting, and confusing. That field is being validated according to `typing.Union[pd.PositiveFloat, typing.Literal[False]]`, so I think `override_structure_dl=0` is being coerced into `False` in the validation. which disables the override structure. This _might_ actually be OK behavior? What do you think @tylerflex? To be fair we could also just change this to `pd.NonNegativeFloat`, I think this will lead to exactly the same behavior. `False` would be interpreted as `0` in that case.

> `learning_rate=0` passed the validation. Should we validate it for >0?

Absolutely, fixed.

tylerflex commented:
> That field is being validated according to typing.Union[pd.PositiveFloat, typing.Literal[False]], so I think override_structure_dl=0 is being coerced into False in the validation. which disables the override structure. This might actually be OK behavior? What do you think @tylerflex? To be fair we could also just change this to pd.NonNegativeFloat, I think this will lead to exactly the same behavior. False would be interpreted as 0 in that case.

seems like expected behavior to me, although a bit confusing. I'd think from the user perspective, changing to `NonNegativeFloat` would be exactly the same? So maybe we can just leave it as is?

tylerflex commented:
something to note, we should update this to use the new initialization spec 

https://github.com/flexcompute/tidy3d/blob/ce648575a8c218e4a6a259ec406bd09446b40493/tidy3d/plugins/invdes/optimizer.py#L184

yaugenst-flex commented:
Ok, I think everything is addressed now except for the issue with the duplicate adjoint field monitors, which I think we need to investigate further, but it is not related to validation.

tylerflex commented on tidy3d/plugins/expressions/metrics.py:
```diff
@@ -40,35 +40,35 @@ class ModeAmp(Metric):
 
     monitor_name: str
-    freqs: FreqArray
+    f: FreqArray
```
can f take a float? I feel this would be potentially better for usability (similar to xarray coords). could add it later I suppose, but I feel making users do `f=[f]` when we only support a single frequency feels like it could be improved.

tylerflex replied:
Also, could we make it optional? if `None` we just isel the single frequency?

yaugenst-flex replied:
pretty sure float works fine, pydantic will coerce into list on validation

yaugenst-flex replied:
if we make it optional we can’t validate for single freq here. do you want that validation to happen somewhere else? in design would be possible along with the other post init stuff. but to me doesnt seem like the right place

yaugenst-flex replied:
```python
m = ModeAmp(f=1, monitor_name="")
print(m.f)  # prints [1.]
```

tylerflex replied:
we can just raise an error when the metric is evaluated? and then the post init validation check should catch it automatically?

yaugenst-flex replied:
no strong opinion there other than the fact that the validator would once again not show what actually failed in the log 😆 (except if you look at the stack trace)
but either way it's a bit of a crutch until we have multifrequency figured out for this, so it doesn't really matter

tylerflex reviewed (changes requested):
Looks good overall, I still take some issue with the Metric.f thing, see comments. Basically I feel like it should operate roughly like xarray, so it's easy to take my existing postprocessing `sel()` code and transport it here, specitically
* allow single `f` value, not just array.
* if `f` is left out it automatically is selected (only works if single `f` in dataset because we dont support arrays).

Thoughts?

yaugenst-flex committed:
feat: Support for arbitrary padding sizes in autograd plugin

yaugenst-flex committed:
feat: Method for extracting sub-components from nested expressions

yaugenst-flex committed:
Additional constraints and validators for `invdes` plugin

 - Constraints for design parameters to be in range [0, 1]
 - Constraints on optimizer, penalty, transformation arguments such as
   `beta`, `eta`, `learning_rate`
 - Validation for forward and gradient evaluation of both
   transformations and penalties
 - Validation for metrics included in `Design` objects
 - Renamed `Metric.freqs` --> `Metric.f`
 - Make `Metric.f` optional
